app/main.py

```diff
@@ -1,44 +1,49 @@
+import asyncio
 import time
 
 from iot.devices import HueLightDevice, SmartSpeakerDevice, SmartToiletDevice
 from iot.message import Message, MessageType
 from iot.service import IOTService
 
 
-def main() -> None:
-    # create an IOT service
+async def main() -> None:
     service = IOTService()
 
-    # create and register a few devices
     hue_light = HueLightDevice()
     speaker = SmartSpeakerDevice()
     toilet = SmartToiletDevice()
-    hue_light_id = service.register_device(hue_light)
-    speaker_id = service.register_device(speaker)
-    toilet_id = service.register_device(toilet)
 
-    # create a few programs
+    device_task1 = asyncio.create_task(service.register_device(hue_light))
+    device_task2 = asyncio.create_task(service.register_device(speaker))
+    device_task3 = asyncio.create_task(service.register_device(toilet))
+
+    await asyncio.gather(device_task1, device_task2, device_task3)
+
     wake_up_program = [
-        Message(hue_light_id, MessageType.SWITCH_ON),
-        Message(speaker_id, MessageType.SWITCH_ON),
-        Message(speaker_id, MessageType.PLAY_SONG, "Rick Astley - Never Gonna Give You Up"),
+        Message(device_task1.result(), MessageType.SWITCH_ON),
+        Message(device_task2.result(), MessageType.SWITCH_ON),
+        Message(
+            device_task3.result(),
+            MessageType.PLAY_SONG,
+            "Rick Astley - Never Gonna Give You Up"
+        ),
     ]
 
     sleep_program = [
-        Message(hue_light_id, MessageType.SWITCH_OFF),
-        Message(speaker_id, MessageType.SWITCH_OFF),
-        Message(toilet_id, MessageType.FLUSH),
-        Message(toilet_id, MessageType.CLEAN),
+        Message(device_task1.result(), MessageType.SWITCH_OFF),
+        Message(device_task2.result(), MessageType.SWITCH_OFF),
+        Message(device_task3.result(), MessageType.FLUSH),
+        Message(device_task3.result(), MessageType.CLEAN),
```

Comment:
Using `device_task.result()` assumes that the tasks have completed successfully and returned valid device IDs. Ensure that the tasks are awaited and completed before accessing their results to avoid potential runtime errors.